pkgs/os-specific/linux/kernel/common-config.nix

```diff
@@ -1,726 +1,704 @@
-  RUNTIME_TESTING_MENU? n
-  # For older kernels, painstakingly disable each symbol.
-  ${optionalString (versionOlder version "4.16") ''
-    ARM_KPROBES_TEST? n
-    ASYNC_RAID6_TEST? n
-    ATOMIC64_SELFTEST? n
-    BACKTRACE_SELF_TEST? n
-    INTERVAL_TREE_TEST? n
-    PERCPU_TEST? n
-    RBTREE_TEST? n
-    TEST_BITMAP? n
-    TEST_BPF? n
-    TEST_FIRMWARE? n
-    TEST_HASH? n
-    TEST_HEXDUMP? n
-    TEST_KMOD? n
-    TEST_KSTRTOX? n
-    TEST_LIST_SORT? n
-    TEST_LKM? n
-    TEST_PARMAN? n
-    TEST_PRINTF? n
-    TEST_RHASHTABLE? n
-    TEST_SORT? n
-    TEST_STATIC_KEYS? n
-    TEST_STRING_HELPERS? n
-    TEST_UDELAY? n
-    TEST_USER_COPY? n
-    TEST_UUID? n
-  ''}
-
-  CRC32_SELFTEST? n
-  CRYPTO_TEST? n
-  ${optionalString (versionOlder version "4.18") ''
-    DRM_DEBUG_MM_SELFTEST? n
-    LNET_SELFTEST? n
-  ''}
-  EFI_TEST? n
-  GLOB_SELFTEST? n
-  LOCK_TORTURE_TEST? n
-  MTD_TESTS? n
-  NOTIFIER_ERROR_INJECTION? n
-  RCU_PERF_TEST? n
-  RCU_TORTURE_TEST? n
-  TEST_ASYNC_DRIVER_PROBE? n
-  WW_MUTEX_SELFTEST? n
-  XZ_DEC_TEST? n
-
-  ${optionalString (features.criu or false)  ''
-    EXPERT y
-    CHECKPOINT_RESTORE y
-  ''}
-
-  ${optionalString ((features.criu or false) && (features.criu_revert_expert or true))
-    # Revert some changes, introduced by EXPERT, when necessary for criu
-  ''
-    RFKILL_INPUT? y
-    HID_PICOLCD_FB? y
-    HID_PICOLCD_BACKLIGHT? y
-    HID_PICOLCD_LCD? y
-    HID_PICOLCD_LEDS? y
-    HID_PICOLCD_CIR? y
-    DEBUG_MEMORY_INIT? y
-  ''}
-
-  ${extraConfig}
-''
+with import ../../../../lib/kernel.nix { inherit (stdenv) lib; inherit version; };
+
+let
+  # temporary hack
+  grsecurity = false;
+  xen_dom0 = false;
+
+
+  # TODO configuration items have to be part of subattrs. Remove this constraint
+  flattenKConf =  nested: mapAttrs (_: head) (zipAttrs (attrValues nested));
+
+  options = {
+
+    debug = {
+      DEBUG_INFO                = if (features.debug or false) then yes else no;
+      DEBUG_KERNEL              = yes;
+      DEBUG_DEVRES              = no;
+      TIMER_STATS               = whenOlder "4.11" yes;
+      DEBUG_NX_TEST             = whenOlder "4.11" no;
+      CPU_NOTIFIER_ERROR_INJECT = whenOlder "4.4" (option no);
+      DEBUG_STACK_USAGE         = no;
+      DEBUG_STACKOVERFLOW       = when (!grsecurity) no;
+      RCU_TORTURE_TEST          = no;
+      SCHEDSTATS                = no;
+      DETECT_HUNG_TASK          = yes;
+      CRASH_DUMP                = option no;
+      # Easier debugging of NFS issues.
+      SUNRPC_DEBUG              = whenAtLeast "3.4" yes;
+    };
+
+    power-management = {
+      PM_RUNTIME                       = whenOlder "3.19" yes;
+      PM_ADVANCED_DEBUG                = yes;
+      X86_INTEL_LPSS                   = whenAtLeast "3.11" yes;
+      X86_INTEL_PSTATE                 = whenAtLeast "3.10" yes;
+      INTEL_IDLE                       = yes;
+      CPU_FREQ_DEFAULT_GOV_PERFORMANCE = yes;
```

Comment:
We may want to add the other governors as options here

Reply:
Let's do such changes separately from the structured-options refactoring; that way it is much easier to verify there weren't unintended changes.

Reply:
agreed, I had to remove some copumpkin CONFIG_ additions that would generate errors on recent kernels.